### ISSUE TITLE
Double processing of messages, recovery by Durability Agent (possible race condition)

### ISSUE THREAD

**Issue by afmuller** [JasperFx/wolverine#587]
Hi Jeremy
I’ve upgraded to Wolverine 1.8.2. Herewith feedback.

Both nodes, at startup, report the node number in the below message on start up.
Wolverine assigned node id for envelope persistence is 144/145

I’ve have full logging turned on on my Postgresql installation. While looking at the logs of the SQL statements, I have noted the below:

1. Income messages are inserted in the wolverine_incoming_envelopes table owner_id = 0
2. After the message has been processed, status is set to ‘Handled’. For most messages, the owner_id is never changed (between the INSERT and the UPDATE setting the status to ‘Handled’)
3. The Durability Agent runs on a schedule and selects all messages in the wolverine_incoming_envelopes table with owner_id = 0 and status = 'Incoming'

**_I suspect the issue to be a race condition._** 
The Durability Agent, running on the schedule (selecting all messages in the wolverine_incoming_envelopes table with owner_id = 0 and status = 'Incoming') picks up records just in inserted (but before being set to 'Handled').

This appears to be confirmed by the logs of a 'double processed messages'. See SQL logs and application logs below:

**2023-10-10 17:00:17.785 SAST** (new message is inserted into the database)
```
INSERT INTO wolverine_incoming_envelopes (id, body, execution_time, message_type, owner_id, received_at, status)
VALUES ($1, $2, $3, $4, $5, $6, $7)
RETURNING attempts","parameters: $1 = '018b1a1a-5eef-4b49-ba79-d25e341e3117', $2 = '<removed>', $3 = NULL, $4 = 'Aquelle.Ops.Application.Features.Jobs.Commands.NewJobExecutionRecordStarted', $5 = '0', $6 = 'local://aquelle.ops.application.features.jobs.commands.newjobexecutionrecordstarted/', $7 = 'Incoming'
```

**2023-10-10 17:00:18.133** (Durability agent runs on schedule - picks up the above message - confimed with timestamp on scheenshot of application logs)
`select received_at, count(*) from public.wolverine_incoming_envelopes where status = 'Incoming' and owner_id = 0 group by received_at`

**2023-10-10 17:00:18.155 SAST** (I only noted this SQL statement for the double-processed message. So I suspect it is the recovered message setting the owner)
`update public.wolverine_incoming_envelopes set owner_id = $1 where id = ANY($2)","parameters: $1 = '136', $2 = '{018b1a1a-5eef-4b49-ba79-d25e341e3117}'`

**2023-10-10 17:00:19.034 SAST** (message set to handled)
`update public.wolverine_incoming_envelopes set status = 'Handled', keep_until = $1 where id = $2","parameters: $1 = '2023-10-10 17:05:19.033664+02', $2 = '018b1a1a-5eef-4b49-ba79-d25e341e3117'`

**2023-10-10 17:00:19.038 SAST** (message is set to handled **again**)
`update public.wolverine_incoming_envelopes set status = 'Handled', keep_until = $1 where id = $2","parameters: $1 = '2023-10-10 17:05:19.037644+02', $2 = '018b1a1a-5eef-4b49-ba79-d25e341e3117'`

See corresponding screenshot of logs:

![image](https://github.com/JasperFx/wolverine/assets/15875383/ed8de232-5b69-43dc-9c4c-1c7404173c88)

**Comment by jeremydmiller** [comment]
This:

```
INSERT INTO wolverine_incoming_envelopes (id, body, execution_time, message_type, owner_id, received_at, status)
VALUES ($1, $2, $3, $4, $5, $6, $7)
RETURNING attempts","parameters: $1 = '018b1a1a-5eef-4b49-ba79-d25e341e3117', $2 = '<removed>', $3 = NULL, $4 = 'Aquelle.Ops.Application.Features.Jobs.Commands.NewJobExecutionRecordStarted', $5 = '0', $6 = 'local://aquelle.ops.application.features.jobs.commands.newjobexecutionrecordstarted/', $7 = 'Incoming'
```

Should **not** be possible. That's the root cause of all the errors. The owner_id should never be zero in that case and I don't yet understand how that's happening.

**Comment by jeremydmiller** [comment]
Do you have any kind of `IHostedService` registered that is sending messages on startup maybe?

**Comment by afmuller** [comment]
Hi Jeremy

To try and reproduce the problem, I have created a very simple application which I have uploaded onto my Github.
[https://github.com/afmuller/WolverineTestWebApp](url)

It is a simple Web API application. 
If you want to run it, you'll need to provide correct database credentials in appsettings.json. 

If you do an HTTP POST to the /Company endpoint, it will insert or update the company in the database (postgres). 
As part of the handler, an outgoing message of CompanyCreated or CompanyUpdated is returned. 

In the CompanyCreated or CompanyUpdated handlers I have inserted a Task.Delay to allow me to see the database record status. 
See below - while the transaction is executing, the Status = Incoming and the owner_id = 0.

![image](https://github.com/JasperFx/wolverine/assets/15875383/f219d0ec-7fe8-4001-a98d-cc00ba5bcaec)

SQL logs:

INSERT with owner_id = 0:
```
INSERT INTO wolverine_incoming_envelopes (id, body, execution_time, message_type, owner_id, received_at, status)
VALUES ($1, $2, $3, $4, $5, $6, $7)
RETURNING attempts","parameters: $1 = '018b1ec5-f8e1-4d21-a20a-54816b144175', $2 = '<removed>', $3 = NULL, $4 = 'WebApplication1.Companies.CompanyUpdated', $5 = '0', $6 = 'local://webapplication1.companies.companyupdated/', $7 = 'Incoming'
```

UPDATE to Handled:
`update public.wolverine_incoming_envelopes set status = 'Handled', keep_until = $1 where id = $2","parameters: $1 = '2023-10-11 14:51:32.868796+02', $2 = '018b1ec5-f8e1-4d21-a20a-54816b144175'`

I havn't reproduced the duplicate messages in this project (as I don't see the Durability Agent calls to the database, for some reason), however, it does show records in the wolverine_incoming_envelopes table with owner_id = 0.

**Comment by afmuller** [comment]
For your information, I've converted and pushed a 'sqlserver' branch to execute against SQL server in stead of postgresql - just to check if it works correctly there. 

It does the same on SQL server. 
See screenshot below:

![image](https://github.com/JasperFx/wolverine/assets/15875383/ab40bdf8-d3fb-4254-bb63-2a39b6723064)

Durability Agent running to recover the message, message is executed twice.
![image](https://github.com/JasperFx/wolverine/assets/15875383/70cd8989-ed70-4b54-bf34-1cd0eb09b260)

**Comment by jeremydmiller** [comment]
@afmuller Easy money with the reproduction steps you gave me. It was in the Wolverine + EF Core mechanics, and nothing to do with any kind of race condition like I'd theorized. Couple line code fix. Only impacted usages *with* EF Core envelope transactions. 

The fix will be in 1.8.3 within the hour.
